src/tree/mod.rs

```diff
@@ -737,122 +747,70 @@ impl RatchetTree {
-            let node = &tree.nodes[index.as_usize()];
-            match node.node_type {
-                NodeType::Leaf => {
-                    let leaf_node_hash = LeafNodeHashInput::new(&index, &node.key_package);
-                    leaf_node_hash.hash(ciphersuite)
+
+    /// Verify the parent hash of a tree node. Returns `true` if the parent
+    /// hash has successfully been verified and `false` otherwise.
+    pub fn verify_parent_hash(&self, index: NodeIndex, node: &Node) -> bool {
+        let left = treemath::left(index).unwrap();
+        let right = treemath::right(index, self.leaf_count()).unwrap();
+        // Unwrapping here is safe, because we know it is a full parent node
+        let parent_hash_field = &node.parent_hash().unwrap();
+        // Current hash with right child resolution
+        let current_hash_right = ParentHashInput::new(&self, index, right, parent_hash_field)
+            // It is ok to use `unwrap()` here, since we can be sure the node is not
+            // blank
+            .unwrap()
+            .hash(&self.ciphersuite);
+
+        // Verify the left child first
+        if let Some(left_parent_hash_field) = self.nodes[left].parent_hash() {
+            if left_parent_hash_field == current_hash_right {
+                return true;
+            }
+        } else {
+            // Current hash with left child resolution
+            let current_hash_left = ParentHashInput::new(&self, index, left, parent_hash_field)
+                // It is ok to use `unwrap()` here, since we can be sure the node is not
+                // blank
+                .unwrap()
+                .hash(&self.ciphersuite);
+            // Then verify the right child
+            if let Some(right_parent_hash_field) = self.nodes[right].parent_hash() {
+                if right_parent_hash_field == current_hash_left {
+                    return true;
                 }
-                NodeType::Parent => {
-                    let left = treemath::left(index)
-                        .expect("node_hash: Error when computing left child of node.");
-                    let left_hash = node_hash(ciphersuite, tree, left);
-                    let right = treemath::right(index, tree.leaf_count())
-                        .expect("node_hash: Error when computing left child of node.");
-                    let right_hash = node_hash(ciphersuite, tree, right);
-                    let parent_node_hash = ParentNodeHashInput::new(
-                        index.as_u32(),
-                        &node.node,
-                        &left_hash,
-                        &right_hash,
-                    );
-                    parent_node_hash.hash(ciphersuite)
+            } else {
+                // Iterate over children until we find a node that is neither blank
+                // nor a leaf node
+                let mut child = right;
+                while child.is_parent() && !self.nodes[child].is_blank() {
```

Comment:
Shouldn't the condition here be exactly the inverse? Currently, we're iterating until we find a node that is either not a parent or blank.

Reply:
The entire function needs more tests, this should probably be one test case.

Reply:
I took another stab at this and refactored in a way that is as close as possible to the spec.

The test we have in place right now is not good enough to cover all code paths. Before I write (pretty complex) test cases to cover everything I need to check with Joel, if this really needs to be so complex. The iteration down the tree seems redundant to me. In the meantime I created #313.